craft_providers/bases/instance_config.py

```diff
@@ -118,6 +118,26 @@ def save(
             file_mode="0644",
         )
 
+    @classmethod
+    def _update_nested_dictionaries(
+        cls, config_data: Dict[str, Any], new_data: Dict[str, Any]
+    ) -> Dict[str, Any]:
```

Comment:
It's weird that it changes the received object and then returns it. I would expect that it returns the result if the original is not modified, or viceversa.

Also, consider isolating this code (which is great) in a regular function outside the class.

Reply:
I moved the code outside the class in the same file.

I agree that modifying a python dictionary and returning it is somewhat odd.  Based on your comment, I tried conditionally returning the result when the original dictionary is modified.  Between the extra code and the extra type checking to satisfy the linters, the function got more complex than it needed to be.  So I reverted.

I checked on stack overflow for a better solution, most solutions seem to be following a similar paradigm.

Reply:
Do not change the behaviour conditionally. Or *always* change the received dict and return `None`, or *always* don't touch the received dict and return the new object. I'd go for the former (and note the behaviour in the docstring).